serverkeyapi/internal/api.go

```diff
@@ -2,20 +2,52 @@ package internal
 
 import (
 	"context"
+	"crypto/ed25519"
+	"encoding/json"
 	"fmt"
 	"time"
 
+	"github.com/matrix-org/dendrite/internal/config"
 	"github.com/matrix-org/dendrite/serverkeyapi/api"
 	"github.com/matrix-org/gomatrixserverlib"
+	"github.com/sirupsen/logrus"
 )
 
 type ServerKeyAPI struct {
 	api.ServerKeyInternalAPI
 
+	Cfg        *config.Dendrite
 	OurKeyRing gomatrixserverlib.KeyRing
 	FedClient  *gomatrixserverlib.FederationClient
 }
 
+func (s *ServerKeyAPI) QueryLocalKeys(ctx context.Context, request *api.QueryLocalKeysRequest, response *api.QueryLocalKeysResponse) error {
+	publicKey := s.Cfg.Matrix.PrivateKey.Public().(ed25519.PublicKey)
```

Comment:
I don't know why this is A Thing. We pass the `config` around all components, and if they need the local key then they can just grab it? Why put it here, this seems like more work?

Reply:
Let's put it another way: I would rather have a 'startup time' dep (needing the key in a component's config) than a runtime dep (needing server key API).

Reply:
It seemed like the sensible thing to do, given that I'm not really sure why the federation API would own control of which keys we hand out to other servers otherwise.

It also means that the server key API tests are testing the *same* key retrieval behaviour that `/matrix/key/v2/server` does in the real world, which means that we haven't duplicated code and some off-hand change in either the server key API or the federation API won't render the test useless and/or break federation without us realising it.

Reply:
We rely on sytests to test endpoints to ensure we don't break federation without us realising it. There's currently a fair amount of different code paths for local vs non-local keys so I can't really see how it saves much on code duplication either. For example, `handleLocalKeys` vs `handleDatabaseKeys`, and several places in `handleFetcherKeys` where we conditionally check if it's our keys to make sure we don't `storeResults`.

Reply:
I'll put it back, but you might understand my hesitation to rely much on Sytest here given that it hasn't picked up on a single issue with any of the key behaviours and federation has still died nightly in many cases as a result.